spec/webauthn/public_key_credential_with_assertion_spec.rb

```diff
@@ -0,0 +1,300 @@
+          )
+        end
+
+        it "works" do
+          expect(
+            public_key_credential.verify(
+              challenge,
+              public_key: credential_public_key,
+              sign_count: credential_sign_count
+            )
+          ).to be_truthy
+
+          expect(public_key_credential.client_extension_outputs).to be_empty
+        end
+      end
+
+      context "is received" do
+        let(:public_key_credential) do
+          WebAuthn::PublicKeyCredentialWithAssertion.new(
+            type: credential_type,
+            id: credential_id,
+            raw_id: credential_raw_id,
+            client_extension_outputs: { "appid" => "true" },
+            response: assertion_response
+          )
+        end
+
+        it "works" do
+          expect(
+            public_key_credential.verify(
+              challenge,
+              public_key: credential_public_key,
+              sign_count: credential_sign_count
+            )
+          ).to be_truthy
+
+          expect(public_key_credential.client_extension_outputs).to eq({ "appid" => "true" })
+        end
+      end
+    end
+
+    context "when authentication extension input" do
+      context "is not received" do
+        let(:assertion_response) do
+          response = client.get(challenge: raw_challenge, extensions: nil)["response"]
+
+          WebAuthn::AuthenticatorAssertionResponse.new(
+            authenticator_data: response["authenticatorData"],
+            client_data_json: response["clientDataJSON"],
+            signature: response["signature"]
+          )
+        end
+
+        it "works" do
+          expect(
+            public_key_credential.verify(
+              challenge,
+              public_key: credential_public_key,
+              sign_count: credential_sign_count
+            )
+          ).to be_truthy
+
+          expect(public_key_credential.authenticator_extension_outputs).to be_nil
+        end
+      end
+
+      context "is received" do
+        let(:assertion_response) do
+          response = client.get(
+            challenge: raw_challenge,
+            extensions: { "txAuthSimple" => "Could you please verify yourself?" }
+          )["response"]
+
+          WebAuthn::AuthenticatorAssertionResponse.new(
+            authenticator_data: response["authenticatorData"],
+            client_data_json: response["clientDataJSON"],
+            signature: response["signature"]
+          )
+        end
+
+        it "works" do
+          expect(
+            public_key_credential.verify(
+              challenge,
+              public_key: credential_public_key,
+              sign_count: credential_sign_count
+            )
+          ).to be_truthy
+
+          expect(public_key_credential.authenticator_extension_outputs)
+            .to eq({ "txAuthSimple" => "Could you please verify yourself?" })
+        end
+      end
+    end
+
+    context "when verifying a migrated U2F credential" do
```

Comment:
Nice!

I think it would be great to  also have `it`s testing when the verification correctly fails if response is not valid.
These test cases are as important as the ones that test it works when all valid, we don't want false positives :-)

Reply:
Good call!